File: App/BNKScopeField/Model/ResourceEngine.swift
import Foundation
import Observation
import BNKKit

/// Lists objects of one kind, and holds what a detail view needs.
@Observable
@MainActor
final class ResourceEngine {
    private(set) var objects: [RawObject] = []
    private(set) var namespaces: [String] = []
    private(set) var loading = false
    private(set) var failure: String?

    var kind: ResourceKind = ResourceKind.all[0]
    var namespace: String?
    var query = ""

    private(set) var events: [K8s.Event] = []
    private(set) var loadingEvents = false

    var visible: [RawObject] {
        let needle = query.trimmingCharacters(in: .whitespaces).lowercased()
        guard !needle.isEmpty else { return objects }
        return objects.filter {
            $0.name.lowercased().contains(needle) || ($0.namespace ?? "").lowercased().contains(needle)
        }
    }

    func loadNamespaces(_ cluster: ManagedCluster) async {
        guard let client = try? cluster.client() else { return }
        namespaces = (try? await client.namespaces()) ?? []
    }

    func load(_ cluster: ManagedCluster) async {
        guard let client = try? cluster.client() else { return }
        loading = true
        failure = nil
        do {
            objects = try await client.list(kind, namespace: kind.namespaced ? namespace : nil)
                .sorted { ($0.namespace ?? "", $0.name) < ($1.namespace ?? "", $1.name) }
        } catch {
            objects = []
            failure = TelemetryEngine.brief(error)
        }
        loading = false
    }

    /// What has been said about one object.
    ///
    /// The reason to open a pod at all is usually that something is wrong with
    /// it, and the events are where the cluster says what.
    func loadEvents(for object: RawObject, cluster: ManagedCluster) async {
        events = []
        guard let namespace = object.namespace, let client = try? cluster.client() else { return }
        loadingEvents = true
        events = (try? await client.events(about: object.name, namespace: namespace)) ?? []
        loadingEvents = false
    }
}

/// The one-line summary a row shows, per kind.
///
/// Written per kind rather than generically because a useful summary is
/// different every time: a pod's is its readiness, a service's is its address, a
/// node's is its version.
enum ResourceSummary {
    static func line(for object: RawObject, kind: ResourceKind) -> (text: String, tone: SummaryTone) {
        switch kind.plural {
        case "pods":
            let statuses = object.array("status", "containerStatuses")
            let ready = statuses.filter { ($0["ready"] as? Bool) == true }.count
            let restarts = statuses.compactMap { ($0["restartCount"] as? NSNumber)?.intValue }.max() ?? 0
            let phase = object.string("status", "phase") ?? "?"
            let node = object.string("spec", "nodeName") ?? "—"
            let text = "\(ready)/\(statuses.count) ready · \(phase)"
                + (restarts > 0 ? " · \(restarts) restarts" : "") + " · \(node)"
            let healthy = phase == "Running" && ready == statuses.count && !statuses.isEmpty
            return (text, phase == "Succeeded" ? .neutral : (healthy ? .good : .bad))

        case "deployments", "statefulsets":
            let ready = object.int("status", "readyReplicas") ?? 0
            let wanted = object.int("spec", "replicas") ?? 0
            return ("\(ready)/\(wanted) ready", ready == wanted ? .good : .bad)

        case "daemonsets":
            let ready = object.int("status", "numberReady") ?? 0
            let wanted = object.int("status", "desiredNumberScheduled") ?? 0
            return ("\(ready)/\(wanted) ready", ready == wanted ? .good : .bad)

        case "services":
            let type = object.string("spec", "type") ?? "ClusterIP"
            let ip = object.string("spec", "clusterIP") ?? "—"
            let ports = object.array("spec", "ports")
                .compactMap { ($0["port"] as? NSNumber)?.stringValue }.joined(separator: ",")
            return ("\(type) · \(ip)\(ports.isEmpty ? "" : " · \(ports)")", .neutral)

        case "configmaps":
            let keys = (object.json["data"] as? [String: Any])?.count ?? 0
            return ("\(keys) key\(keys == 1 ? "" : "s")", .neutral)

        case "events":
            let type = object.string("type") ?? "?"
            let reason = object.string("reason") ?? ""
            let about = object.string("involvedObject", "name") ?? ""
            return ("\(reason) · \(about)", type == "Warning" ? .bad : .neutral)

        case "nodes":
            let conditions = object.array("status", "conditions")
            let ready = conditions.contains { ($0["type"] as? String) == "Ready" && ($0["status"] as? String) == "True" }
            let version = object.string("status", "nodeInfo", "kubeletVersion") ?? "?"
            let arch = object.string("status", "nodeInfo", "architecture") ?? ""
            return ("\(ready ? "Ready" : "NotReady") · \(version) · \(arch)", ready ? .good : .bad)

        default:
            return ("", .neutral)
        }
    }
}

enum SummaryTone { case good, bad, neutral }

File: dart/bnk_engines/lib/src/resource_engine.dart
import 'package:bnk_kit/bnk_kit.dart';

import 'brief.dart';
import 'cluster_store.dart';
import 'observable.dart';

enum SummaryTone { good, bad, neutral }

/// Lists objects of one kind, and holds what a detail view needs.
class ResourceEngine extends Observable {
  List<RawObject> _objects = const [];
  List<String> _namespaces = const [];
  bool _loading = false;
  String? _failure;
  List<Event> _events = const [];
  bool _loadingEvents = false;

  ResourceKind _kind = ResourceKind.all[0];
  String? _namespace;
  String _query = '';

  List<RawObject> get objects => _objects;
  List<String> get namespaces => _namespaces;
  bool get loading => _loading;
  String? get failure => _failure;
  List<Event> get events => _events;
  bool get loadingEvents => _loadingEvents;

  ResourceKind get kind => _kind;
  set kind(ResourceKind v) {
    _kind = v;
    notify();
  }

  String? get namespace => _namespace;
  set namespace(String? v) {
    _namespace = v;
    notify();
  }

  String get query => _query;
  set query(String v) {
    _query = v;
    notify();
  }

  List<RawObject> get visible {
    final needle = _query.trim().toLowerCase();
    if (needle.isEmpty) return _objects;
    return [
      for (final o in _objects)
        if (o.name.toLowerCase().contains(needle) || (o.namespace ?? '').toLowerCase().contains(needle)) o
    ];
  }

  Future<void> loadNamespaces(ManagedCluster cluster) async {
    final client = cluster.clientOrNull;
    if (client == null) return;
    try {
      _namespaces = await client.namespaces();
    } catch (_) {
      _namespaces = const [];
    }
    notify();
  }

  Future<void> load(ManagedCluster cluster) async {
    final client = cluster.clientOrNull;
    if (client == null) return;
    _loading = true;
    _failure = null;
    notify();
    try {
      final list = await client.list(_kind, namespace: _kind.namespaced ? _namespace : null);
      list.sort((a, b) {
        final ns = (a.namespace ?? '').compareTo(b.namespace ?? '');
        return ns != 0 ? ns : a.name.compareTo(b.name);
      });
      _objects = list;
    } catch (e) {
      _objects = const [];
      _failure = brief(e);
    }
    _loading = false;
    notify();
  }

  /// What has been said about one object. The reason to open a pod at all
  /// is usually that something is wrong with it, and the events are where
  /// the cluster says what.
  Future<void> loadEvents(RawObject object, ManagedCluster cluster) async {
    _events = const [];
    final namespace = object.namespace;
    final client = cluster.clientOrNull;
    if (namespace == null || client == null) {
      notify();
      return;
    }
    _loadingEvents = true;
    notify();
    try {
      _events = await client.events(about: object.name, namespace: namespace);
    } catch (_) {
      _events = const [];
    }
    _loadingEvents = false;
    notify();
  }
}

/// The one-line summary a row shows, per kind.
///
/// Written per kind rather than generically because a useful summary is
/// different every time: a pod's is its readiness, a service's is its
/// address, a node's is its version.
class ResourceSummary {
  static ({String text, SummaryTone tone}) line(RawObject object, ResourceKind kind) {
    switch (kind.plural) {
      case 'pods':
        final statuses = object.array(['status', 'containerStatuses']);
        final ready = statuses.where((s) => s['ready'] == true).length;
        var restarts = 0;
        for (final s in statuses) {
          final r = s['restartCount'];
          if (r is num && r > restarts) restarts = r.toInt();
        }
        final phase = object.string(['status', 'phase']) ?? '?';
        final node = object.string(['spec', 'nodeName']) ?? '—';
        final text = '$ready/${statuses.length} ready · $phase'
            '${restarts > 0 ? ' · $restarts restarts' : ''} · $node';
        final healthy = phase == 'Running' && ready == statuses.length && statuses.isNotEmpty;
        return (text: text, tone: phase == 'Succeeded' ? SummaryTone.neutral : (healthy ? SummaryTone.good : SummaryTone.bad));

      case 'deployments' || 'statefulsets':
        final ready = object.integer(['status', 'readyReplicas']) ?? 0;
        final wanted = object.integer(['spec', 'replicas']) ?? 0;
        return (text: '$ready/$wanted ready', tone: ready == wanted ? SummaryTone.good : SummaryTone.bad);

      case 'daemonsets':
        final ready = object.integer(['status', 'numberReady']) ?? 0;
        final wanted = object.integer(['status', 'desiredNumberScheduled']) ?? 0;
        return (text: '$ready/$wanted ready', tone: ready == wanted ? SummaryTone.good : SummaryTone.bad);

      case 'services':
        final type = object.string(['spec', 'type']) ?? 'ClusterIP';
        final ip = object.string(['spec', 'clusterIP']) ?? '—';
        final ports = [for (final p in object.array(['spec', 'ports'])) if (p['port'] != null) '${p['port']}'].join(',');
        return (text: '$type · $ip${ports.isEmpty ? '' : ' · $ports'}', tone: SummaryTone.neutral);

      case 'configmaps':
        final data = object.json['data'];
        final keys = data is Map ? data.length : 0;
        return (text: '$keys key${keys == 1 ? '' : 's'}', tone: SummaryTone.neutral);

      case 'events':
        final type = object.string(['type']) ?? '?';
        final reason = object.string(['reason']) ?? '';
        final about = object.string(['involvedObject', 'name']) ?? '';
        return (text: '$reason · $about', tone: type == 'Warning' ? SummaryTone.bad : SummaryTone.neutral);

      case 'nodes':
        final ready = object.array(['status', 'conditions']).any((c) => c['type'] == 'Ready' && c['status'] == 'True');
        final version = object.string(['status', 'nodeInfo', 'kubeletVersion']) ?? '?';
        final arch = object.string(['status', 'nodeInfo', 'architecture']) ?? '';
        return (text: '${ready ? 'Ready' : 'NotReady'} · $version · $arch', tone: ready ? SummaryTone.good : SummaryTone.bad);

      default:
        return (text: '', tone: SummaryTone.neutral);
    }
  }
}
